### ISSUE TITLE
Support multiple reporters

### ISSUE THREAD

**Issue by mgrebenets** [realm/SwiftLint#1227]
Hi

I want to convert existing JSON report to HTML format instead of re-running SwiftLint again on the whole source code. Is there a way to do so?

It would be nice to be able to run SwiftLint once and then convert reports from one format to another.

It would be probably even better to be able to run SwiftLint once with multiple `--reporter` options and get multiple reports in one run.

**Comment by marcelofabri** [comment]
> I want to convert existing JSON report to HTML format instead of re-running SwiftLint again on the whole source code. Is there a way to do so?

Currently, no. But it shouldn't very hard to do so. I just think this doesn't belong to SwiftLint itself.

> It would be probably even better to be able to run SwiftLint once with multiple --reporter options and get multiple reports in one run.

I agree that this is a better solution. I will change the issue title to reflect this request.

**Comment by amlcurran** [comment]
Hi, as commented on the previous issue (#1162), I'd really like this to be supported.

If the Swiftlint team agree with the idea, I'd be really happy to implement it and submit a PR.

**Comment by jpsim** [comment]
Could you share a design with us before implementing it? What would usage look like, how would it coexist with the current approach?

**Comment by amlcurran** [comment]
Sure.

I would imagine it would be relatively simple - instead of the current single string for reporter, it could either be a comma-separated string or a YAML sequence:

```yaml
reporter: "xcode,junit"
```

or

```yaml
reporter:
  - xcode
  - junit
```

Either way, the current version would be maintained for compatibility.

Codewise, it only looks like the changes would be required would be in [`LintCommand`](https://github.com/realm/SwiftLint/blob/master/Source/swiftlint/Commands/LintCommand.swift#L53). I'd probably attempt something like the [composite patten](https://en.wikipedia.org/wiki/Composite_pattern). I see that there is a `isRealtime` bool in the `Reporter` protocol - I'd have to consider whether this has any impact, or you could help me out there?

**Comment by jpsim** [comment]
Ok, but where would it output? Reporters currently log to stdout, so what would multiple reporters output to? Files? How would you specify those? If not files, what else?

**Comment by amlcurran** [comment]
Good point! I'll think about how could work - and how we would want it to
work from our main use case - and get back to you.

**Comment by marcelofabri** [comment]
Just for reference, this is how rubocop works:

> You can enable multiple formatters at the same time by specifying `-f/--format` multiple times. The `-o/--out` option applies to the previously specified `-f/--format`, or the default progress format if no `-f/--format` is specified before the `-o/--out` option.

```
# Simple format to $stdout.
$ rubocop --format simple

# Progress (default) format to the file result.txt.
$ rubocop --out result.txt

# Both progress and offense count formats to $stdout.
# The offense count formatter outputs only the final summary,
# so you'll mostly see the outputs from the progress formatter,
# and at the end the offense count summary will be outputted.
$ rubocop --format progress --format offenses

# Progress format to $stdout, and JSON format to the file rubocop.json.
$ rubocop --format progress --format json --out rubocop.json
#         ~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~
#                 |               |_______________|
#              $stdout

# Progress format to result.txt, and simple format to $stdout.
$ rubocop --output result.txt --format simple
#         ~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~
#                  |                 |
#           default format        $stdout
```

(from [their docs](http://rubocop.readthedocs.io/en/latest/formatters/))

**Comment by jpsim** [comment]
Definitely can't do something like that out-of-the-box using [Commandant](https://github.com/Carthage/Commandant).

**Comment by iliaskarim** [comment]
This still seems like a good idea. It may not be the most elegant solution, however, this is what I can come up with.

Generate an intermediate representation of the output. Then run a swiftlint command to translate that intermediate format into the desired reporter output. That would not require linting all the files again, since the same intermediate file could be used in multiple swiftlint commands to output into different reporters. Naively speaking, a reporter format like the JSON format ought to be translatable into another format without having to design a new intermediate format.

**Comment by jpsim** [comment]
This should definitely be easier to build now that we migrated from Commandant to swift-argument-parser.

Although you should note that because SwiftLint caches its results, running swiftlint twice with different reporters should be pretty fast considering the files don't need to be linted again if they haven't changed.
